numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -1088,6 +1099,16 @@ PyArray_PromoteTypes(PyArray_Descr *type1, PyArray_Descr *type2)
         return type1;
     }
 
+    if (PyDataType_HASFIELDS(type1) || PyDataType_HASFIELDS(type2)) {
+        if (!PyArray_EquivTypes(type1, type2)) {
+            PyErr_SetString(PyExc_TypeError,
+                    "cannot promote structured types with non-equivalent fields");
+            return NULL;
+        }
+        Py_INCREF(type1);
+        return type1;
```

Comment:
Can you add a test for this behavior? In that if two dtypes with different padding are specified, the first/last one is returned by identity?

Reply:
Actually the way I wrote it using "equivalent_fields"  this rejects promotion for dtypes with different padding/offsets. Probably that's safer anyway so no-one gets their padding changed unexpectedly.

Added a test for that case.